kaggle_environments/envs/lux_ai_2022/luxai2022/actions.py

```diff
@@ -0,0 +1,331 @@
+from collections import defaultdict
+from dataclasses import dataclass
+from enum import Enum
+from typing import Callable
+
+import numpy as np
+from typing import TYPE_CHECKING
+from luxai2022.config import EnvConfig
+if TYPE_CHECKING:
+    from luxai2022.factory import Factory
+    from luxai2022.state import State
+from luxai2022.map.position import Position
+
+
+import luxai2022.unit as luxai_unit
+
+
+# (0 = move, 1 = transfer X amount of R, 2 = pickup X amount of R, 3 = dig, 4 = self destruct, 5 = recharge X, 6 = repeat)
+class Action:
+    def __init__(self, act_type: str) -> None:
+        self.act_type = act_type
+        self.repeat = False
+
+    def state_dict():
+        raise NotImplementedError("")
+
+
+class FactoryBuildAction(Action):
+    def __init__(self, unit_type: luxai_unit.UnitType) -> None:
+        super().__init__("factory_build")
+        self.unit_type = unit_type
+
+    def state_dict(self):
+        if self.unit_type == luxai_unit.UnitType.LIGHT:
+            return 0
+        return 1
+
+
+class FactoryWaterAction(Action):
+    def __init__(self) -> None:
+        super().__init__("factory_water")
+        self.water_cost = None
+
+    def state_dict():
+        return 2
+
+
+class MoveAction(Action):
+    def __init__(self, move_dir: int, dist: int = 1, repeat=False) -> None:
+        super().__init__("move")
+        # a[1] = direction (0 = center, 1 = up, 2 = right, 3 = down, 4 = left)
+        self.move_dir = move_dir
+        self.dist = dist
+        self.repeat = repeat
```

Comment:
can we not have this associated with the action queue instead of on each action?

Reply:
A little bit complicated to make it associated with the entire action queue instead of per action. This is mostly for people using deep learning approaches. The additional granularity also gives more flexibility as well.

Associating with the action queue instead of each action makes the action space slightly more complicated as it's not a simple matrix of size (K, 5) anymore where each action is 5 dimensions and the queue is of length K. It would be a dictionary with an additional boolean attribute which might not work as nicely with some RL libraries (not all support dictionary based actions easily).

Also I theorize this may make the credit assignment problem (of state-action values) a bit more difficult in RL, especially for people who might consider auto-regressive generation of actions.